libp2p/pubsub/pubsub.py

```diff
@@ -505,18 +530,39 @@ def handle_subscription(
 
         # TODO: Check if the `from` is in the blacklist. If yes, reject.
 
-        # TODO: Check if signing is required and if so signature should be attached.
-
         # If the message is processed before, return(i.e., don't further process the message).
         if self._is_msg_seen(msg):
             return
 
-        # TODO: - Validate the message. If failed, reject it.
-        # Validate the signature of the message
-        # FIXME: `signature_validator` is currently a stub.
-        if not signature_validator(msg.key, msg.SerializeToString()):
-            logger.debug("Signature validation failed for msg: %s", msg)
-            return
+        # Check if signing is required and if so signature should be attached.
+        if self.strict_signing:
+            if msg.signature == b"":
+                logger.debug("Reject because no signature attached for msg: %s", msg)
+                return
+            # Validate if message sender matches message signer,
+            # i.e., check if `msg.key` matches `msg.from_id`
+            msg_pubkey = deserialize_public_key(msg.key)
+            if ID.from_pubkey(msg_pubkey) != msg.from_id:
+                logger.debug(
+                    "Reject because signing key does not match sender ID for msg: %s",
+                    msg,
+                )
+                return
+            # Validate the signature of the message
+            # First, construct the original payload that's signed by 'msg.key'
+            msg_without_key_sig = rpc_pb2.Message(
+                data=msg.data,
+                topicIDs=msg.topicIDs,
+                from_id=msg.from_id,
+                seqno=msg.seqno,
+            )
+            payload = (
+                PUBSUB_SIGNING_PREFIX.encode() + msg_without_key_sig.SerializeToString()
+            )
```

Comment:
Would it be better to move these verification logics to `signature_validator` and change its function signature to `signature_validator(msg: rpc_pb2.Message)`? IMO it seems to be more instinct.